# KART - K-way Admixture and Relationship on Tractor
### Tractor without Hail


## Introduction

This script was implemeneted based on the model published in Nature Genetics Atkinson et al (https://www.nature.com/articles/s41588-020-00766-y). In our tests we used teh RFMix 1, with the output converted to RFMix 2 model using the script RFMix1ToRFMix2.py.

This pipeline has, as new features: 

- The model is easily aplicable to a N-way admixed population. 

- The model uses Generalized linear mixed model (package GMMAT on R) to uses a kinship matrix and the whole data from imputed data

- The author of the methodology recomends to run the Local Ancestry using the original file, so imputed SNPs between two windows inherit the ancestry of the nearest window


### Aplicability to a N-way admixed population

Our script is easily aplicable to N parental because the use of the equation

![Equation](./Equations/CodeCogsEqn.png)

Which sum means summation the H_j represents the number haplotypes for the ancestry j (all ancestries except one), the D_i represents the dosage for the ancestry i (All ancestries) and the "... + b_wX_w" means the other covariates. 

To do this our model uses the files generated by step 1 from the original pipeline (https://github.com/eatkinson/Tractor/wiki/Step-1:-Recovering-tracts-(Optional)) except for bcftools annotate. 

### Generalized linear mixed model (package GMMAT on R) to uses a kinship matrix and the whole data from imputed data

Using the "Unkinked" files our pipeline creates a covariative file for each SNP and run SNP by SNP a glmm.wald (slow, we try address this using multiprocessing). We also create a GDS file based on the VCF provided by the user, but we make P copies (P = number of processors to be used) of this GDS file because this file accepts one access per time, so each processor will have theis own GDS file.

To be faster, we do not exclude the temporary files while run the regression, but the final file will be one with all tested SNPs (<outputPrefix>_AllWald.txt). We also provide a flag (-s) to allow the user change the model to select the interest covariatives (the script will not set  the variables from Tractor automatically, the user will need to put it)
  

## How to use

Our pipeline uses R with the libraries SeqArray and GMMAT installed, bcftools and python 3.

We recommend the use of kinship matrix (you can convert from PLINK or KING using the scripts present in this GitHub).

Example of command line:

```
python3.8 main.py -c Covar.txt -v /home/USER/TopMed/WithoutHWE/chr22.dose.unkinked_Covar_MAF.vcf.gz -n 3 -m /home/USER/TopMed/WithoutHWE/Tractor/RFMixV1_chr22_AllSets.Unkinked.msp.tsv -i ID -p DISEASE -f /home/USER/TopMed/WithoutHWE/Tractor/newModel/Teste/ -o Try3 -t 48 -k PLINK_LARGE.table
```

```
Tractor without Hail

optional arguments:
  -h, --help            show this help message and exit

Required arguments related to genetic files:
  -m MSPPREFIX, --mspPrefix MSPPREFIX
                        MSP prefix (without .msp.fb) file from RFMix (preferred the unkinked)
  -v VCF, --vcf VCF     VCF file (preferred the unkinked)
  -n NUMANCESTRY, --numAncestry NUMANCESTRY
                        Number of ancestries in MSP file

Required arguments related to covar files:
  -c COVAR, --covar COVAR
                        Table with the covariatives
  -i ID, --id ID        Name of the column with individual ID on covar file
  -p PHENOTYPE, --phenotype PHENOTYPE
                        Name of the column with phenotype

Required arguments related to output:
  -o OUTPUT, --output OUTPUT
                        Output prefix
  -f FOLDER, --folder FOLDER
                        Folder to store the results

Optional arguments to model:
  -k KINSHIP, --kinship KINSHIP
                        File with kinship coefficient
  -s STATISTICALMODEL, --statisticalModel STATISTICALMODEL
                        Statistical Model. If not provided, we will include all covariatives
  -T, --Tractor         If the user select the flag -s, this flag insert the Tractor covariates (Default True)

Optional arguments related to system:
  -t THREADS, --threads THREADS
                        Number of threads to run (default = 1)
  -b BCFTOOLS, --bcftools BCFTOOLS
                        Path for the BCFTOOLS (default bcftools)
  -R RSCRIPT, --Rscript RSCRIPT
                        Path for the Rscript with SeqArray and GMMAT installed (default Rscript)
  -e, --exclude         Set to exclude temporary files (default = False)
  -H, --Header          Set to not output the header in all p-values merged (default = True, ie, output has a header)

Optional arguments related to dosage:
  -D DOSAGEFIELD, --dosageField DOSAGEFIELD
                        Select the field to extract the dosage value (default: GT)
  -S SEPARATORDOSAGEFIELD, --separatorDosageField SEPARATORDOSAGEFIELD
                        Separator from select dosage field (default | for GT)
  -P POSITIONDOSAGEFIELD, --positionDosageField POSITIONDOSAGEFIELD
                        Position form the values splited on dosage field (default "0 1" for GT)

Optional arguments related to local ancestry:
  -F, --fowardBackward  Use the posterior probability from Foward backward to count the dosage per ancestry (default False)
```

 #### Acknowledgment
  
Developer: Thiago Peixoto Leal, PhD (thpeixotol@hotmail.com)

Contributors: Mateus Henrique Gouveia, PhD

Special thanks: Elizabeth Atkinson, PhD
